File: ios/VllowDugout/Models/Drill.swift
import Foundation

nonisolated struct Drill: Identifiable, Hashable, Sendable {
    let id: UUID
    let title: String
    let subtitle: String
    let category: DrillCategory
    let difficulty: Difficulty
    let durationSeconds: Int
    let targetHits: Int
    let coinReward: Int
    let imageURL: String
    let isBookmarked: Bool

    init(
        id: UUID = UUID(),
        title: String,
        subtitle: String,
        category: DrillCategory,
        difficulty: Difficulty,
        durationSeconds: Int,
        targetHits: Int,
        coinReward: Int,
        imageURL: String = "",
        isBookmarked: Bool = false
    ) {
        self.id = id
        self.title = title
        self.subtitle = subtitle
        self.category = category
        self.difficulty = difficulty
        self.durationSeconds = durationSeconds
        self.targetHits = targetHits
        self.coinReward = coinReward
        self.imageURL = imageURL
        self.isBookmarked = isBookmarked
    }

    var durationFormatted: String {
        let minutes = durationSeconds / 60
        let seconds = durationSeconds % 60
        return "\(minutes)M \(seconds)S"
    }
}

nonisolated enum DrillCategory: String, CaseIterable, Sendable {
    case batting = "Batting"
    case bowling = "Bowling"
    case fielding = "Fielding"
    case wicketkeeping = "Wicketkeeping"
    case masterclass = "Masterclass"
}

nonisolated enum Difficulty: String, CaseIterable, Sendable {
    case beginner = "Beginner"
    case intermediate = "Intermediate"
    case pro = "Pro"
    case elite = "Elite"
}

File: flutter_app/lib/models/drill.dart
import 'package:uuid/uuid.dart';

enum DrillCategory {
  batting('Batting'),
  bowling('Bowling'),
  fielding('Fielding'),
  wicketkeeping('Wicketkeeping'),
  masterclass('Masterclass');

  final String displayName;
  const DrillCategory(this.displayName);
}

enum Difficulty {
  beginner('Beginner'),
  intermediate('Intermediate'),
  pro('Pro'),
  elite('Elite');

  final String displayName;
  const Difficulty(this.displayName);
}

class Drill {
  final String id;
  final String title;
  final String subtitle;
  final DrillCategory category;
  final Difficulty difficulty;
  final int durationSeconds;
  final int targetHits;
  final int coinReward;
  final String imageURL;
  final bool isBookmarked;

  Drill({
    String? id,
    required this.title,
    required this.subtitle,
    required this.category,
    required this.difficulty,
    required this.durationSeconds,
    required this.targetHits,
    required this.coinReward,
    this.imageURL = '',
    this.isBookmarked = false,
  }) : id = id ?? const Uuid().v4();

  String get durationFormatted {
    final minutes = durationSeconds ~/ 60;
    final seconds = durationSeconds % 60;
    return '${minutes}M ${seconds}S';
  }

  Drill copyWith({
    String? id,
    String? title,
    String? subtitle,
    DrillCategory? category,
    Difficulty? difficulty,
    int? durationSeconds,
    int? targetHits,
    int? coinReward,
    String? imageURL,
    bool? isBookmarked,
  }) {
    return Drill(
      id: id ?? this.id,
      title: title ?? this.title,
      subtitle: subtitle ?? this.subtitle,
      category: category ?? this.category,
      difficulty: difficulty ?? this.difficulty,
      durationSeconds: durationSeconds ?? this.durationSeconds,
      targetHits: targetHits ?? this.targetHits,
      coinReward: coinReward ?? this.coinReward,
      imageURL: imageURL ?? this.imageURL,
      isBookmarked: isBookmarked ?? this.isBookmarked,
    );
  }

  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is Drill && id == other.id;

  @override
  int get hashCode => id.hashCode;
}
